cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,473 @@
+  * cilium status - overview of authentication status
+    * N/M pods configured for authentication
+    * N authentication sessions with peers
+  * cilium auth list - List all authentication sessions in cilium-agent
+    * Local node
+    * Peer node
+    * Local pod (selector?)
+    * Peer pod (selector?)
+    * Last authentication timestamp
+    * ETA next authentication timestamp
+    * Inbound / Outbound session establishment (i.e. triggered locally or by remote peer)
+    * Status (authenticating / authenticated / failed + err )
+  * cilium bpf auth list - Dump bpf map state for authentication (if applicable)
+* hubble events - Add "authenticated" timestamp to events so that hubble CLI, UI, etc can visualize authentication status of connections between in-cluster peers
+  * Perhaps, "last authentication time" + "next authentication time"?
+  * Was authentication required by the local peer or the remote peer for this traffic?
+  * Discussion needed: SIG-Hubble
+* Metrics
+  * Total authentication sessions in the cilium-agent (inbound vs. outbound)
+  * SPIFFE ID registration time (during CNI ADD)
+  * TLS Session authentication time (during connection establishment) 
+  * New SPIFFE Identities (with expiry age)
+  * Updated SPIFFE Identities (with expiry age)
+  * mTLS Handshake stats (succeeded, failed, reasons, etc.)
+  * Your metrics suggestions here
+
+
+## Impacts / Key Questions
+List crucial impacts and key questions. They likely require discussion and are required to understand the trade-offs of the CFP. During the lifecycle of a CFP, discussion on design aspects can be moved into this section. After reading through this section, it should be possible to understand any potentially negative or controversial impact of this CFP. it should also be possible to derive the key design questions: X vs Y.
+
+### Impact: Increased latency for authenticated sessions
+The current design primarily relies upon dropping the first packet of a connection to trigger authentication. This will increase latency for new connections and will likely cause Cilium to compare poorly with corresponding implementations on a fresh start (e.g. Istio). This will be looked at closely as soon as we have a working implementation, as we intend to launch the feature with as much pre-warming as possible to mitigate or eliminate the packet drops.
+
+### Impact: Authentication connections per node
+Depending on the options for establishing connections between nodes (link) and options for the server to understand which client is establishing an authenticated connection from a peer node (link), connections may be established as:
+* One connection per pair of nodes
+* One connection per pair of namespaces
+* One connection established per pair of deployments/identities establishing data connections, i.e. corresponding to the pod A (node A) -> pod B (node B)
+* Two connections per pair, node A -> pod B (node B) + node B -> pod A (node A)
```

Comment:
nit: I don't know if it's worth updating this, but I think that the CFP has ended up adopting a combination of two of the options here:

```suggestion
Each node will establish multiple connections to other nodes in the cluster. For each unique identity for a local workload that makes outbound connections to another security identity on another node, a new auth session will be initiated. Likewise, the destination will also establish a connection to validate the authenticity of the same inbound connection at the remote node.
```

If we wanted to go whole-hog here then some diagrams could help - cases like:
- One local application connects to a single deployment that has two backends on two remote nodes. This results in four connections between two security identities:
  - Node A -> Node B for identity x->y (At Node A)
  - Node A -> Node B for identity x->y (At Node B)
  - Node A -> Node C for identity x->y (At Node A)
  - Node A -> Node C for identity x->y (At Node C)
- Similarly, for one application that makes calls to two different remote Pods on other nodes with different identities, there would be four connections (even the same peer node) :
  - Node A -> Node B for identity x->y (At Node A) 
  - Node A -> Node B for identity x->y (At Node B) 
  - Node A -> Node B for identity x->z (At Node A) 
  - Node A -> Node B for identity x->z (At Node B) 

We could probably explore even more cases, but this is probably enough info for now to consider the 'back of the hand' math on the impact.

Reply:
^ "At Node ____" above assumes that the specified node is the one initiating the authentication connection.

Reply:
I agree about your suggestion here, but I'd like us to come back and update this section once we've gone through how the current working prototype is doing its thing, then we can think more about the table size and impacts.

Reply:
I logged cilium/cilium/#24617 to cover coming back and checking the anticipated table sizes.